cmd/thanos/rule.go

```diff
@@ -324,11 +393,98 @@ func queryPrometheusInstant(ctx context.Context, addr, query string, t time.Time
 	return vec, nil
 }
 
-func printAlertNotifications(ctx context.Context, expr string, alerts ...*rules.Alert) error {
-	fmt.Fprintf(os.Stdout, "%d alerts for %s\n", len(alerts), expr)
-	for _, a := range alerts {
-		fmt.Fprintf(os.Stdout, "  labels: %s val: %f\n", a.Labels, a.Value)
+type alertmanagerSet struct {
+	resolver *net.Resolver
+	addrs    []string
+	mtx      sync.Mutex
+	current  []*url.URL
+}
+
+func newAlertmanagerSet(addrs []string, resolver *net.Resolver) *alertmanagerSet {
+	if resolver == nil {
+		resolver = net.DefaultResolver
+	}
+	return &alertmanagerSet{
+		resolver: resolver,
+		addrs:    addrs,
 	}
+}
+
+func (s *alertmanagerSet) get() []*url.URL {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.current
+}
+
+const defaultAlertmanagerPort = 9093
+
+func (s *alertmanagerSet) update(ctx context.Context) error {
+	var res []*url.URL
+
+	for _, addr := range s.addrs {
```

Comment:
Would be nice to reuse that inside known-hosts flag too.. but maybe in separate PR

Reply:
Yes, we should reconcile once we've settled for this.